### ISSUE TITLE
Imagenet Training Error

### ISSUE THREAD

**Issue by LukeAI** [AlexeyAB/darknet#3671]
Training as below fails with a mysterious error when it tries to calculate the first validation accuracy
- Using the default cfg/darknet19.cfg (except subdivisions doubled) with normal Imagenet
- compiled with cuda 10.0 CuDNN 7.5.0 cudnn_half
- latest repo. RTX 2080Ti Cuda 10.0 Cudnn 7.5.2 driver 418.56 ubuntu-server 16.04

any ideas? :)

I have double-checked that all the paths are correct.
 ```
$ cat imagenet1k_c.data
classes=1000
train  = /srv/imagenet/imagenet1k.train.list
valid  = /srv/imagenet/inet.val.list
backup = backup
labels = /srv/imagenet/imagenet.labels.list
names  = /srv/imagenet/imagenet.shortnames.list
top=5

$ ./darknet classifier train imagenet1k_c.data cfg/darknet19.cfg -dont_show -mjpeg_port 8090 -topk
...
...
trains as expected for 20000 iterations
...
...
MJPEG-stream sent.
Loaded: 0.000038 seconds

 try to allocate additional workspace_size = 52.43 MB
 CUDA allocate done!
 TOP calculation...
49999: top 1: 0.095800, top 5: 0.239860
 try to allocate additional workspace_size = 518.00 MB
 CUDA allocate done!

 accuracy TOP5 = 0.239860
20018, 2.000: 2.560756, 2.594085 avg, 0.095089 rate, 447.311554 seconds, 2562304 images
 MJPEG-stream sent.
Loaded: 0.000041 seconds

 cuDNN status Error in: file: ./src/convolutional_kernels.cu : () : line: 539 : build time: Jul 24 2019 - 10:53:21
cuDNN Error: CUDNN_STATUS_EXECUTION_FAILED
cuDNN Error: CUDNN_STATUS_EXECUTION_FAILED: File exists
darknet: ./src/utils.c:293: error: Assertion `0' failed.
Aborted (core dumped)
```

EDIT:
I tried rebuilding without CUDNN_HALF just to see what happened
more or less the same although much more accurate:
```
 try to allocate additional workspace_size = 52.43 MB 
 CUDA allocate done! 
 TOP calculation...
17045: top 1: 0.134460, top 5: 0.317160
49999: top 1: 0.133220, top 5: 0.309460
 try to allocate additional workspace_size = 518.00 MB 
 CUDA allocate done! 

 accuracy TOP5 = 0.309460 
20018, 2.000: 2.381682, 2.582306 avg, 0.095089 rate, 448.749573 seconds, 2562304 images
 MJPEG-stream sent. 
Loaded: 0.000048 seconds

 cuDNN status Error in: file: ./src/convolutional_kernels.cu : () : line: 539 : build time: Jul 28 2019 - 16:39:57 
cuDNN Error: CUDNN_STATUS_EXECUTION_FAILED
cuDNN Error: CUDNN_STATUS_EXECUTION_FAILED: File exists
darknet: ./src/utils.c:293: error: Assertion `0' failed.
Aborted (core dumped)
```

**Comment by AlexeyAB** [comment]
* Did you meet this error with other cfg-files or with previous Darknet code?
* Can you check TopK with this dataset and this cfg-file successfully?


> cuDNN Error: CUDNN_STATUS_EXECUTION_FAILED

Usually it meant that something wrong with your cuDNN, CUDA and GPU driver versions.

**Comment by LukeAI** [comment]
I've trained a lot of different object detectors on the same platform without encountering this error, although they would all be with slightly older versions of this repo. - I update whenever I'm trying a new one. I got the same result if I run with cudnn_half=0
I could try checking out an older version and seeing if this is a regression?

**Comment by AlexeyAB** [comment]
Yes, try to use older version.

**Comment by LukeAI** [comment]
I just tried with 9c9344a1ff4541499c8f69ea997be5145b4e1de3 from a few months ago
got same result.

**Comment by LukeAI** [comment]
do you reckon I should just try upgrading cudnn?

**Comment by AlexeyAB** [comment]
Yes, try to reinstall 
GPU drivers
CUDA
cuDNN

**Comment by LukeAI** [comment]
Tried doing the same thing with PJReddie's repo, using the default darknet19.cfg there.
It trained to 600,000 and then gave a bizarrely low top 1 top 5 
`./darknet classifier train imagenet1k_c.data cfg/darknet19.cfg`
`./darknet classifier valid imagenet1k_c.data cfg/darknet19.cfg backup/darknet19.weights`
```
........
/srv/imagenet/labelled/ILSVRC2012_val_00050000.n02437616.JPEG, 185, 0.001030, 0.001131, 
49999: top 1: 0.001000, top 5: 0.005000
```

Any idea what could be up?

**Comment by AlexeyAB** [comment]
Did you  reinstall
GPU drivers
CUDA
cuDNN
?

**Comment by LukeAI** [comment]
I uninstalled cudnn (.deb install) and tried doing a generic linux install (just copying the headers and .so across) for 7.6.2 but then I wondered: does the Make process make assumptions about where cudnn is? If it can't find cudnn will it give useful error messages?

I see this line in Makefile:
 CFLAGS+= -DCUDNN -I/usr/local/cudnn/include
LDFLAGS+= -L/usr/local/cudnn/lib64 -lcudnn
but that's not where I've put cudnn!

**Comment by AlexeyAB** [comment]
> If it can't find cudnn will it give useful error messages?

Yes, if you set CUDNN=1

so try to find all cudnn libraries in your PC

**Comment by LukeAI** [comment]
ok reinstalling cudnn from the generic linux .tgz package has resolved this. Thanks as ever, for the support.

**Comment by LukeAI** [comment]
After training for 800,000 iterations it crashed again in the same way:

 accuracy TOP5 = 0.823580 
880792, 87.999: 1.384623, 1.172708 avg, 0.004083 rate, 453.043304 seconds, 112741376 images
 MJPEG-stream sent. 
Loaded: 0.000028 seconds

 cuDNN status Error in: file: ./src/convolutional_kernels.cu : () : line: 538 : build time: Jul 30 2019 - 20:04:44 
cuDNN Error: CUDNN_STATUS_EXECUTION_FAILED
cuDNN Error: CUDNN_STATUS_EXECUTION_FAILED: Broken pipe
darknet: ./src/utils.c:293: error: Assertion `0' failed.
Aborted (core dumped)

I just restarted training again using the last weights. hopefully it will make it.

**Comment by LukeAI** [comment]
it made it to 900,000 and crashed again :/

**Comment by AlexeyAB** [comment]
Does it happen only with this cfg-file?

Do you observe any strange behavior or too large RAM / VRAM memory consumption?

**Comment by LukeAI** [comment]
I have only tried training a darknet classifier with this .cfg - i've trained many yolos without this issue. I'll try and keep a closer eye on ram / vram over the next couple hours - the crash always happens immediately after calculating validation accuracy as above.

**Comment by AlexeyAB** [comment]
Try to debug, or at least add `printf(" 1 \n");` 
`printf(" 2 \n");` 
`printf(" 3 \n");`
...

at different points in the code to catch the place of error

**Comment by LukeAI** [comment]
The crash happens after some of them, not all of them.

**Comment by LukeAI** [comment]
So I put in some printf("1\n") statements into classifier.c from line 124 to try and work out where things are breaking down:
The error happens some time after printf("2\n") and before printf("6\n")
I don't know if this is useful for narrowing it down much? I'll try and get more fine grained insight.
```
        printf("Loaded_!!: %lf seconds\n", sec(clock()-time));
        time=clock();

        float loss = 0;
#ifdef GPU
        printf("1\n");
        if(ngpus == 1){
            printf("2\n");
            loss = train_network(net, train);
            printf("3\n");
        } else {
            printf("4\n");
            loss = train_networks(nets, ngpus, train, 4);
            printf("5\n");
        }
#else
        loss = train_network(net, train);
#endif
        if(avg_loss == -1 || isnan(avg_loss) || isinf(avg_loss)) avg_loss = loss;
        avg_loss = avg_loss*.9 + loss*.1;
        printf("6\n");

```


```
output:
Loaded_!!: 0.000032 seconds
1
2

 cuDNN status Error in: file: ./src/convolutional_kernels.cu : () : line: 539 : build time: Aug  7 2019 - 10:43:07
```

**Comment by martinraism** [comment]
I am having the exact same error at the exact same line (539 of convolutional_kernels.cu).
I have an application that loads and unloads darknet. When I first load darknet, eveything works smoothly and fine. However, after unloading it and loading it back, at the moment of the first convolution of the network, this happens.
I am using the latest version of darknet from Alexey, cudnn 7.3, cuda 10 (also tried and happening with cuda 10.1) and Nvidia driver version  419.67.

**Comment by LukeAI** [comment]
I added some printf statements after around line 385 in network.c

The error happens in between "a" and "b" - presumably in train_network_datum. I can keep digging...
```
    for(i = 0; i < n; ++i){
        printf("i: %d\n", i);
        get_next_batch(d, batch, i*batch, X, y);
        printf("a");
        net.current_subdivision = i;
        float err = train_network_datum(net, X, y);
        printf("b");
```

**Comment by AlexeyAB** [comment]
@LukeAI 


Try to check in which layer it is happen.
* https://github.com/AlexeyAB/darknet/blob/480b3eccb19ac65508e49fb3116ccda65551bb6e/src/network_kernels.cu#L52-L53
* https://github.com/AlexeyAB/darknet/blob/480b3eccb19ac65508e49fb3116ccda65551bb6e/src/network_kernels.cu#L98-L99

And does it happen during forward or backward ?

**Comment by LukeAI** [comment]
```
forward_network_gpu
i = 0

 cuDNN status Error in: file: ./src/convolutional_kernels.cu : () : line: 539 : build time: Aug  9 2019 - 10:02:04 
```
```

void forward_network_gpu(network net, network_state state)
{
    //cudaDeviceSynchronize();
    printf("forward_network_gpu\n");
    state.workspace = net.workspace;
    int i;
    for(i = 0; i < net.n; ++i){
        printf("i = %d\n", i);
        state.index = i;
        layer l = net.layers[i];

```

**Comment by LukeAI** [comment]
Added some more printf checkpoints:
```
    for(i = 0; i < net.n; ++i){
        printf("i = %d\n", i);
        state.index = i;
        layer l = net.layers[i];

        if(l.delta_gpu && state.train){
            printf("a");
            fill_ongpu(l.outputs * l.batch, 0, l.delta_gpu, 1);
            printf("b");
        }
        //printf("\n layer %d - type: %d - \n", i, l.type);
        //start_timer();
        l.forward_gpu(l, state);
        printf("c");
```


```
tnd_1forward_network_gpu
i = 0
ab
 cuDNN status Error in: file: ./src/convolutional_kernels.cu : () : line: 539 : build time: Aug  9 2019 - 14:52:05
```

**Comment by AlexeyAB** [comment]
@LukeAI There is also should be output - what is the error: `cuDNN Error .... `

----

https://github.com/AlexeyAB/darknet/blob/4c315ea26b56c2bf20ebc240d94386c6e3cc83db/src/dark_cuda.c#L203-L204
https://github.com/AlexeyAB/darknet/blob/4c315ea26b56c2bf20ebc240d94386c6e3cc83db/src/dark_cuda.c#L176-L192

**Comment by AlexeyAB** [comment]
So may be we should use separate instance of network for Top1 calculation for Classifier, as we done it for Detector.

**Comment by LukeAI** [comment]
Same as above:
```
cuDNN status Error in: file: ./src/convolutional_kernels.cu : () : line: 539 : build time: Jul 24 2019 - 10:53:21
cuDNN Error: CUDNN_STATUS_EXECUTION_FAILED
cuDNN Error: CUDNN_STATUS_EXECUTION_FAILED: File exists
darknet: ./src/utils.c:293: error: Assertion `0' failed.
Aborted (core dumped)
```

**Comment by LukeAI** [comment]
> So may be we should use separate instance of network for Top1 calculation for Classifier, as we done it for Detector.
> 
> In any cases this is very strange error.

Perhaps? I have no idea :)
It's strange how it seems to happen at random - usually not, just sometimes.

**Comment by LukeAI** [comment]
As a temp workaround I've stopped using the -topk flag - will this stop the periodic topk calculations? Otherwise I just have to check the training every few hours to see if it has crashed.

**Comment by hooman67** [comment]
I'm facing the exact same issue as @LukeAI.

4cuDNN Error: CUDNN_UNKNOWN_STATUS: Broken pipe
darknet: ./src/utils.c:293: error: Assertion `0' failed.
Aborted (core dumped)

So far, I've tried different datasets, different configs, and different GPUs (on the same machine). I have also tried cuda 9.0, cuda 9.1, cuda 10.0, and cuda 10.1. For each dataset/config the issue is caused the first time the validation accuracy is calculated. Weird thing is that I can train with the same data, config, and cuda version on a different machine with no problem.
